File: backend/app/api/v1/block_chain.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from app.db.base import get_session as get_db
from app.schemas.block_chain import MiningResult, TransactionPoolStatus
from app.services.mining import MiningService

router = APIRouter(prefix="/api/v1/blockchain", tags=["auth"])



@router.post("/mine", response_model=MiningResult)
async def mine_block(
        miner_address: str,
        max_transactions: int = 10,
        db: AsyncSession = Depends(get_db)
):
    """手动挖矿"""
    service = MiningService(db)
    result = await service.mine_block(miner_address, max_transactions)

    if not result.success:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="挖矿失败，可能没有待处理交易或系统正在挖矿中"
        )

    return result



@router.get("/transaction-pool", response_model=TransactionPoolStatus)
async def get_transaction_pool_status(
        db: AsyncSession = Depends(get_db)
):
    """获取交易池状态（基于异步会话）"""
    from app.db.models.block_chain import TransactionPool

    # 统计待处理交易数量
    stmt_count = select(func.count()).select_from(TransactionPool)
    result_count = await db.execute(stmt_count)
    pending_transactions = result_count.scalar() or 0

    # 统计交易总金额
    stmt_total = select(func.sum(TransactionPool.amount))
    result_total = await db.execute(stmt_total)
    total_value = result_total.scalar() or 0

    # 统计平均 gas 费
    stmt_avg = select(func.avg(TransactionPool.gas_fee))
    result_avg = await db.execute(stmt_avg)
    avg_gas_fee = result_avg.scalar() or 0

    return TransactionPoolStatus(
        pending_transactions=pending_transactions,
        total_value=total_value,
        average_gas_fee=avg_gas_fee,
    )


# 新增：交易池明细列表接口
@router.get("/transaction-pool/list")
async def get_transaction_pool_list(
        page: int = 1,
        limit: int = 10,
        db: AsyncSession = Depends(get_db)
):
    """
    获取交易池明细列表（分页）
    返回结构：{"items": [...], "total": int}
    """
    from app.db.models.block_chain import TransactionPool

    # 统计总数
    stmt_count = select(func.count()).select_from(TransactionPool)
    result_count = await db.execute(stmt_count)
    total = result_count.scalar() or 0

    # 分页查询
    stmt_list = (
        select(TransactionPool)
        .order_by(TransactionPool.created_at.desc())
        .offset((page - 1) * limit)
        .limit(limit)
    )
    result_list = await db.execute(stmt_list)
    items = result_list.scalars().all()

    # 序列化为普通 dict，便于前端使用
    serialized_items = []
    for tx in items:
        serialized_items.append({
            "id": tx.id,
            "transaction_hash": tx.transaction_hash,
            "from_address": tx.from_address,
            "to_address": tx.to_address,
            "amount": tx.amount,
            "gas_fee": tx.gas_fee,
            "data": tx.data,
            "priority_score": tx.priority_score,
            "created_at": tx.created_at.isoformat() if tx.created_at else None,
        })

    return {
        "items": serialized_items,
        "total": total,
    }


@router.get("/mining/statistics")
async def get_mining_statistics(
        db: AsyncSession = Depends(get_db)
):
    """获取挖矿统计信息（异步版，直接基于区块和交易表统计）"""
    from app.db.models.block_chain import Block, Transaction

    # 区块总数
    stmt_block_count = select(func.count()).select_from(Block)
    result_block_count = await db.execute(stmt_block_count)
    total_blocks = result_block_count.scalar() or 0

    # 交易总数
    stmt_tx_count = select(func.count()).select_from(Transaction)
    result_tx_count = await db.execute(stmt_tx_count)
    total_transactions = result_tx_count.scalar() or 0

    # 最新区块信息
    stmt_last_block = select(Block).order_by(Block.block_number.desc()).limit(1)
    result_last_block = await db.execute(stmt_last_block)
    last_block = result_last_block.scalars().first()

    latest_block_number = last_block.block_number if last_block else None
    latest_block_hash = last_block.block_hash if last_block else None

    return {
        "total_blocks": total_blocks,
        "total_transactions": total_transactions,
        "latest_block_number": latest_block_number,
        "latest_block_hash": latest_block_hash,
    }


@router.get("/chain-info")
async def get_chain_info(
        db: AsyncSession = Depends(get_db)
):
    """
    获取区块链综合信息：
    - total_blocks: 区块总数
    - height: 当前区块高度（total_blocks - 1，若有创世块）
    - total_transactions: 链上交易总数
    - pending_pool_size: 交易池中待处理交易数量
    - latest_block: 最新区块的关键信息
    - chain_valid: 简单链校验结果（当前版本始终返回 True）
    """
    from app.db.models.block_chain import Block, Transaction, TransactionPool

    # 区块总数
    stmt_block_count = select(func.count()).select_from(Block)
    result_block_count = await db.execute(stmt_block_count)
    total_blocks = result_block_count.scalar() or 0

    # 交易总数
    stmt_tx_count = select(func.count()).select_from(Transaction)
    result_tx_count = await db.execute(stmt_tx_count)
    total_transactions = result_tx_count.scalar() or 0

    # 交易池待处理数量
    stmt_pool_count = select(func.count()).select_from(TransactionPool)
    result_pool_count = await db.execute(stmt_pool_count)
    pending_pool_size = result_pool_count.scalar() or 0

    # 最新区块信息
    stmt_last_block = select(Block).order_by(Block.block_number.desc()).limit(1)
    result_last_block = await db.execute(stmt_last_block)
    last_block = result_last_block.scalars().first()

    latest_block = None
    if last_block:
        latest_block = {
            "block_number": last_block.block_number,
            "block_hash": last_block.block_hash,
            "previous_hash": last_block.previous_hash,
            "transactions_count": last_block.transaction_count,
            "timestamp": last_block.timestamp.isoformat() if last_block.timestamp else None,
            "miner_address": last_block.miner_address,
        }

    # 当前高度：如果有创世块（通常 block_number 从 0 开始），高度为 total_blocks - 1，否则为 0
    if total_blocks > 0:
        height = total_blocks - 1
    else:
        height = 0

    # 当前版本不做复杂链校验，先返回 True，后续可扩展为逐块校验 previous_hash / merkle_root 等
    chain_valid = True

    return {
        "total_blocks": total_blocks,
        "height": height,
        "total_transactions": total_transactions,
        "pending_pool_size": pending_pool_size,
        "latest_block": latest_block,
        "chain_valid": chain_valid,
    }


@router.get("/info")
async def get_blockchain_info(
        db: AsyncSession = Depends(get_db)
):
    """
    兼容前端使用的 /info 路由，内部复用 /chain-info 的实现
    """
    return await get_chain_info(db)


@router.post("/mining/start")
async def start_auto_mining(
        miner_address: str,
        interval: int = 30,
        db: AsyncSession = Depends(get_db)
):
    """启动自动挖矿（当前版本不支持，提示使用手动挖矿）"""
    raise HTTPException(
        status_code=status.HTTP_400_BAD_REQUEST,
        detail="当前版本不支持自动挖矿，请使用 /api/v1/blockchain/mine 接口进行手动挖矿",
    )


@router.get("/blocks")
async def get_blocks(
        page: int = 1,
        limit: int = 10,
        db: AsyncSession = Depends(get_db)
):
    """分页获取链上区块"""
    from app.db.models.block_chain import Block

    # 统计总数
    stmt_count = select(func.count()).select_from(Block)
    result_count = await db.execute(stmt_count)
    total = result_count.scalar() or 0

    # 分页查询
    stmt_list = (
        select(Block)
        .order_by(Block.block_number.desc())
        .offset((page - 1) * limit)
        .limit(limit)
    )
    result_list = await db.execute(stmt_list)
    items = result_list.scalars().all()

    # 序列化
    serialized = []
    for b in items:
        serialized.append({
            "block_number": b.block_number,
            "block_hash": b.block_hash,
            "previous_hash": b.previous_hash,
            "transactions_count": b.transaction_count,
            "timestamp": b.timestamp.isoformat() if b.timestamp else None,
            "miner_address": b.miner_address
        })

    return {
        "items": serialized,
        "total": total
    }

File: fronted/src/api/Blockchain.ts
// frontend/src/api/Blockchain.ts
// 区块链相关接口封装，统一从这里调用

import service from './http'
/**
 * 链状态信息，从 /api/v1/blockchain/info 获取
 */
export interface LatestBlockInfo {
  block_number: number
  block_hash: string
  previous_hash: string
  transactions_count: number
  timestamp: string | null
  miner_address: string
}

export interface ChainInfo {
  total_blocks: number          // 区块总数
  height: number                // 当前区块高度
  total_transactions: number    // 链上总交易（捐赠）笔数
  pending_pool_size: number     // 交易池待处理数量
  latest_block: LatestBlockInfo | null
  chain_valid: boolean          // 链是否通过校验
}

/**
 * 区块简要信息，用于列表展示，对应 /api/v1/blockchain/blocks
 */
export interface BlockSummary {
  height: number
  block_hash: string
  prev_hash: string | null
  timestamp: string
  tx_count: number
  total_amount_in_block: number
  project_ids?: string | null
  meta?: string | null
}

/**
 * 区块列表返回结构
 */
export interface BlocksResponse {
  total: number
  offset: number
  limit: number
  items: BlockSummary[]
}

/**
 * 区块内单笔交易信息，对应后端 ChainTransactionInfo
 */
export interface ChainTransactionInfo {
  id: number
  project_id: number
  donor_username: string | null
  amount: number
  remark: string | null
  tx_hash: string
  tx_index: number
  timestamp: string
  external_donate_id: number | null
}

/**
 * 区块详情，对应 /api/v1/blockchain/blocks/{height}
 */
export interface BlockDetail {
  height: number
  block_hash: string
  prev_hash: string | null
  timestamp: string
  tx_count: number
  total_amount_in_block: number
  meta?: string | null
  txs: ChainTransactionInfo[]
}

/**
 * 挖矿结果，对应 /api/v1/blockchain/mine-pending
 */
export interface MinePendingResult {
  message: string
  block_index: number
  block_hash: string
  tx_count: number
}

/**
 * 校验链结果，对应 /api/v1/blockchain/validate
 */
export interface ValidateChainResult {
  valid: boolean
  message: string
}

/**
 * 获取链整体状态：当前高度、总笔数等
 * GET /api/v1/blockchain/info
 */
export const getChainInfo = () =>
  service.get<ChainInfo>('/api/v1/blockchain/info').then(r => r.data)

/**
 * 分页获取区块列表
 * GET /api/v1/blockchain/blocks?offset=&limit=
 */
export const getBlocks = (params: { offset?: number; limit?: number } = {}) =>
  service.get('/api/v1/blockchain/blocks', { params }).then(r => r.data)

/**
 * 获取指定高度的区块详情
 * GET /api/v1/blockchain/blocks/{height}
 */
export const getBlockDetail = (index: number) =>
  service.get(`/api/v1/blockchain/blocks/${index}`).then(r => r.data)

/**
 * 触发一次“打包待处理交易并挖矿”
 * POST /api/v1/blockchain/mine-pending
 */
export const minePendingBlocks = () =>
  service.post('/api/v1/blockchain/mine-pending').then(r => r.data)

/**
 * 校验当前链是否有效
 * POST /api/v1/blockchain/validate
 */
export const validateBlockchain = () =>
  service.post('/api/v1/blockchain/validate').then(r => r.data)